### DISCUSSION TITLE
Replacing `#to_graphql` in tests

### DISCUSSION THREAD

**Discussion by slimdave** [rmosolgo/graphql-ruby#3977] [Q&A] [answered]
We're currently on 1.13.10 and preparing for the big 2+ upgrade.

In our tests, we habitually send `to_graphql` to a `Types::BaseInputObject` class before calling `new` on it, although my recollection of _why_ we do that is foggy. 😬

(Was it required in order to instantiate the Singleton? Certainly `Types::Search::WorkSearchAttributes` doesn't seem to respond to `new` until we do)
```
  context "with searches" do
    let(:work_search) do
      Types::Search::WorkSearchAttributes.to_graphql

      Types::Search::WorkSearchAttributes.new(
        {
          subtitleCont:    "subtitle contents",
          subtitleEq:      "subtitle is equal to",
          subtitlePresent: true,
          titleCont:       "title contents",
          titleEq:         "title is equal to",
          idEq:            876,
          after:           543
        },
        context:       OpenStruct.new(current_client: client),
        defaults_used: Set.new
      )
    end
```
I'm wondering what the approved method would be for achieving this aim? Or are we doing things completely wrong here?

**Comment by rmosolgo** [comment, marked as answer]
> _why_ we do that

Yep, this isn't the first time I've heard of it! (Even the graphql-ruby tests did it.) In short, when a legacy Input Object definition was created, it also created a custom `GraphQL::Query::Arguments` subclass which contained some behavior pertaining to the input object's specific arguments. 

Inside `InputObject#initialize`, it would do different kinds of initialization given different inputs. Legacy-style inputs would still work _if_ the custom Arguments subclass was also present. So probably, when migrating from `.define`-based definitions to class-based definitions, adding `.to_graphql` (which, under the hood, prepared that specialized Arguments subclass) made the previous initialization code still work. 

Here are couple examples from the graphql-ruby tests for how to migrate this code: https://github.com/rmosolgo/graphql-ruby/blob/2b35a5d55c1f9889a91f56b313eddfe45ee08bc5/spec/graphql/schema/input_object_spec.rb#L627-L638

In short, pass `nil` as the first positional argument (this is used for legacy arguments), then pass the hash of arguments as `ruby_kwargs: ...` instead. If it doesn't work out of the box, try substituting camel-cased keys for underscore-cased keys (for example, `subtitleEq` => `subtitle_eq`). If it still doesn't work, please share the error or behavior you encounter when you try it!

**Reply by slimdave** [reply]
Oh, that has very nearly got me there!

Where we have:
```ruby
module Types
  module Search
    class ProductSearchAttributes < Types::BaseInputObject
      description "Attributes for searching for products"

      argument  :after,
                Integer,
                description: "Where the product's internal ID is greater than the value",
                required:    false,
                as:          :id_gt
```
... I was testing with ...
```ruby
      Types::Search::ProductSearchAttributes.to_graphql

      Types::Search::ProductSearchAttributes.new(
        {
          after:                   987_654,
```
... but when supplying `ruby_kwargs` it looks as if the test on an argument with `as:` needs to reference (in this example) `id_gt` rather then the argument `after`?

So

```ruby
      Types::Search::ProductSearchAttributes.new(
        nil,
        ruby_kwargs:
          {
            id_gt:                   987_654,
```

... (and hence needing to use underscore-case rather than camel-case)?
```

**Reply by rmosolgo** [reply]
Yeah, I'd believe it! IIRC `as: ...` is applied when turning GraphQL inputs into Ruby keyword arguments. So that'd make sense you'd have to substitute those keys for plain ruby initialization, too.

**Reply by slimdave** [reply]
Makes sense, thanks Robert.
